pkg/status/combinedstatus-resolution.go

```diff
@@ -0,0 +1,333 @@
+				statusCollectorName, err)) // TODO: handle
+			return false, nil
+		}
+
+		updated = updated || changed
+	}
+
+	return updated, nil
+}
+
+// evaluateWorkStatusAgainstStatusCollector evaluates the workstatus against
+// the statuscollector clauses and caches the evaluations. If the workstatus
+// does not match the filter, no other clause is evaluated.
+// The function returns true if an evaluation is updated.
+// If any evaluation fails, the function returns an error.
+// The function assumes that the caller holds a lock over the combinedstatus
+// resolution.
+func evaluateWorkStatusAgainstStatusCollectorLocked(celEvaluator *celEvaluator, workStatusWECName string,
+	workStatusContent *runtime.RawExtension, scData *statusCollectorData) (bool, error) {
+	wsData, exists := scData.workStatusRefToData[workStatusWECName]
+
+	// evaluate filter to determine if the workstatus is relevant
+	if scData.Filter != nil {
+		eval, err := celEvaluator.Evaluate(*scData.Filter, workStatusContent)
+		if err != nil {
+			return false, fmt.Errorf("failed to evaluate filter expression: %w", err)
+		}
+
+		if eval.Type().TypeName() != "bool" {
+			return false, fmt.Errorf("filter expression must evaluate to a boolean")
+		}
+
+		if !eval.Value().(bool) { // workstatus is not relevant
+			if exists { // remove the workstatus data if it exists
+				delete(scData.workStatusRefToData, workStatusWECName)
+				return true, nil
+			}
+
+			return false, nil
+		}
+	}
+
+	if !exists {
+		wsData = &workStatusData{}
+		scData.workStatusRefToData[workStatusWECName] = wsData
+	}
+	updated := false
+
+	// evaluate select
+	selectEvals := make(map[string]ref.Val)
+	for _, selectNamedExp := range scData.Select {
+		eval, err := celEvaluator.Evaluate(selectNamedExp.Def, workStatusContent)
+		if err != nil {
+			return false, fmt.Errorf("failed to evaluate select expression: %w", err)
+		}
+
+		updated = updated || eval.Value() != wsData.selectEval[selectNamedExp.Name]
+		selectEvals[selectNamedExp.Name] = eval
+	}
+
+	// evaluate groupBy
+	groupByEvals := make(map[string]ref.Val)
+	for _, groupByNamedExp := range scData.GroupBy {
+		eval, err := celEvaluator.Evaluate(groupByNamedExp.Def, workStatusContent)
+		if err != nil {
+			return false, fmt.Errorf("failed to evaluate groupBy expression: %w", err)
+		}
+
+		updated = updated || eval.Value() != wsData.groupByEval[groupByNamedExp.Name]
+		groupByEvals[groupByNamedExp.Name] = eval
+	}
+
+	// evaluate combinedFields
+	combinedFieldEvals := make(map[string]ref.Val)
+	for _, combinedFieldNamedAgg := range scData.CombinedFields {
+		if combinedFieldNamedAgg.Type == v1alpha1.AggregatorTypeCount {
+			// count does not require a subject - mark the evaluation with the type
+			currentEval, exists := wsData.combinedFieldsEval[combinedFieldNamedAgg.Name]
+			updated = updated || !exists || currentEval != nil
+
+			combinedFieldEvals[combinedFieldNamedAgg.Name] = nil
```

Comment:
Is it even necessary to have an entry in this map in this case?

Reply:
If the name exists and it is a nil (and not the case of nil because the key is not mapped) then the type is COUNT. This is needed in order to properly update `updated` upon this kind of change.

Reply:
I am not sure that this is the only way to do it. Not a blocker.